stdlib/Dates/src/io.jl

```diff
@@ -379,8 +394,15 @@ function DateFormat(f::AbstractString, locale::DateLocale=ENGLISH)
     prev = ()
     prev_offset = 1
 
-    letters = String(collect(keys(CONVERSION_SPECIFIERS)))
-    for m in eachmatch(Regex("(?<!\\\\)([\\Q$letters\\E])\\1*"), f)
+    # To understand this block, please see the comments attached to the
+    # definitions of DATEFORMAT_REGEX_CACHE and CONVERSION_SPECIFIERS_KEYS_HASH above.
+    conversion_specifiers_keys_hash = hash(keys(CONVERSION_SPECIFIERS))
+    if conversion_specifiers_keys_hash != CONVERSION_SPECIFIERS_KEYS_HASH[]
```

Comment:
Are there a y concurrency concerns here? Maybe a lock is needed?

Reply:
Hypothetically all of these objects need locks I think, i.e. `CONVERSION_SPECIFIERS`, `CONVERSION_DEFAULTS`, `CONVERSION_TRANSLATIONS`, `DATEFORMAT_REGEX_CACHE`, and `CONVERSION_SPECIFIERS_KEY_HASH`. But addressing that strikes me as beyond the scope of this PR: To do so properly, we would need to introduce an interface that maintains consistency between all of these objects, and introduce appropriate reader/writer locks.

Reply:
My thinking was that it used to be thread safe to create date formats but now it no longer is. I would have thought slapping a lock around the part that reads and updates the cache would make this version as safe as the original one. The cost of the lock should ve insignificant compared to the work of the function. Thoughts.

Reply:
That's a good point. It also occurs to me that this would do the wrong thing in case of a hash collision, though of course that is very unlikely.

Reply:
To check my understanding, your concern is specifically about consistency of and between `DATEFORMAT_REGEX_CACHE` and `CONVERSION_SPECIFIERS_KEYS_HASH` across threads constructing `DateFormat`s, when other threads are simultaneously modifying `CONVERSION_SPECIFIERS`? And not about the existing potential for races when some threads modify `CONVERSION_SPECIFIERS` (and friends) while others construct `DateFormat`s (or perform other operations that consume `CONVERSION_SPECIFIERS` and friends)?

Sure, the former we could mitigate pretty simply via an exclusive lock on `CONVERSION_SPECIFIERS_HASH` / `DATEFORMAT_REGEX_CACHE` (at risk of introducing unnecessary contention) or less simply via a combination of shared and exclusive locks (to avoid introducing unnecessary contention). The latter seems like a broader, preexisting issue.

> It also occurs to me that this would do the wrong thing in case of a hash collision, though of course that is very unlikely.

Both of these concerns seem pretty unlikely to bite in practice :).

Reply:
I've pushed a new commit that introduces an exclusive lock (`DATEFORMAT_REGEX_LOCK`) guarding access to `DATEFORMAT_REGEX_CACHE` and `DATEFORMAT_REGEX_HASH` (née `CONVERSION_SPECIFIERS_KEYS_HASH`, which seemed a bit wordy if more descriptive); hopefully the potential contention isn't so bad. Does that address your concern, or are you also concerned about the preexisting issues?